### ISSUE TITLE
Consider more genericity

### ISSUE THREAD

**Issue by dabrahams** [tevelee/swift-graphs#2]
Years ago I was involved in the [Boost Graph Library](https://www.boost.org/doc/libs/1_87_0/libs/graph/doc/index.html), which applies Stepanov's principles of Generic Programming to the graph domain.  Structurally, there are opportunities for efficiency, composability, and code reuse that your library doesn't take advantage of.  LMK if you are interested in discussing further.

**Comment by tevelee** [comment]
Yes, I am absolutely interested! Is there any specific generalization that you have in mind?

**Comment by dabrahams** [comment]
There are many.  Have you studied the BGL at all?  It's not lovely in all respects (C++  after all; Swift is a much better language for Generic Programming), but its fundamental design is beautiful, and recognizes the way graph algorithms build on one another.  It's really worth your time.  I'm not blowing my own horn—I was only marginally involved.

In case you're not aware, Alex Stepanov's work is about raising algorithms to the highest level of abstraction without loss of efficiency.  In the process, if you do it right, you discover deep truths about the domain you're addressing.  This is one of the few libraries outside the STL that really expresses Alex's methodology accurately (he says so).

For example, the A* and Dijkstra algorithms are fundamentally generalizations of BFS, and that is captured by the corresponding visitors.

The BGL's use of property maps decouples graph representation and storage of information associated with vertices and edges from the graph structure representation.

A simple thing I noticed that could be improved is that, for example, your Dijkstra algorithm creates its own predecessor and distance maps and returns them, when it could instead accept generic maps from the user.  Reasons to do that:
- The caller may already have the storage somewhere ready to use.
- The caller may know of a more efficient representation (e.g. if vertices are all described by a dense series of integers, you can use an array)
- The caller may not care about either piece of information.  I have a search where I only care about predecessors.  Since the algorithm never uses the info, I could pass a dummy distance map that discards the “written” values.

Those are just off the top of my head, from having looked only briefly at your code.  A more thorough comparison of your library with BGL would doubtless uncover more.

There's another problem I'm struggling with that wouldn't be solved by any structure I currently know of, but is very worth considering how to acommodate generically:  I want to run an A* search in a space where the outgoing edges of a vertex depend on previously-discovered vertices, so a memo of that information is needed.  By recording distances and predecessors, the algorithm keeps that memo already, but it isn't part of the graph, so any query for the outgoing edges of a vertex can't be satisfied by the graph.  And you really don't want the graph to have to be mutable.  Support for this kind of memoization is probably applicable to almost any algorithm (I've seen `LazyGraph`; the reference semantics of the stored function probably makes some of this possible, and I *suppose* in Swift if you don't make it `Sendable`, it's still threadsafe, but I am very allergic to reference semantics ;-))

**Comment by tevelee** [comment]
Thank you for the insights, Dave! I wasn’t previously familiar with BGL, but the use of visitors and property maps is fascinating. It’s a wealth of inspiration that I’m starting to explore. I plan to embark on this journey, likely beginning with the graph representations.

**Comment by dabrahams** [comment]
Very glad you're inspired, but please begin with the algorithms.  The main graph representation in the library, `adjacency_list`, is something you can't replicate in Swift short of using macros (because it uses template metaprogramming) and is truly, _truly_ not the point of the library.  If you focus on that you will probably not get most of the benefits.  A big point of the library is that it defines graph *concepts*, which would be expressed by protocols in Swift, and any existing graph data structure can be adapted post-hoc to satisfy those concepts.  There is an [excellent book on the BGL](https://www.google.com/books/edition/The_Boost_Graph_Library/CPi7g1hjyIYC?hl=en&gbpv=0) if you want to know more.

Personally, I would be very happy to have a Swift version of the BGL (with fewer overloads, but otherwise very similar).

**Comment by tevelee** [comment]
Apologies, the stale bot got ahead of me! I'm thrilled to share that the library has come a long way toward BGL's design in the past few releases.

Taking your advice to heart (starting with algorithms and concepts rather than representations) the library now has a layered protocol hierarchy (`IncidenceGraph`, `BidirectionalGraph`, `VertexListGraph`, `EdgeListGraph`, `PropertyGraph`, etc.) that maps closely to BGL's graph concepts. Algorithms are external to graph types and constrained to the minimal protocol they need, so any conforming type works automatically.

Property maps are also decoupled now: Cost and Heuristic let callers supply their own weight/heuristic logic, and the visitor pattern (`DijkstraVisitor`, `DFSVisitor`, etc.) lets callers observe and short-circuit execution without modifying the algorithm.

For `AdjacencyList`, I used Swift generics rather than template metaprogramming macros, so storage backends (CSR, COO, ordered, cached bidirectional) are pluggable via generic parameters rather than macro-selected policy classes. Different in mechanism, but similar in spirit. I also found more idiomatic Swift ways to express property maps. 

There's still room to grow, but I wanted to close the loop on a conversation that shaped a lot of the design. Thank you, Dave!

**Comment by dabrahams** [comment]
I'll have to take a look but I'm very impressed that you went this way!  keep up the great work!
